File: src/daemon.ts
/**
 * MCP-CLI Daemon - Background worker that maintains persistent MCP connections
 * 
 * This is spawned as a detached process and manages a Unix socket for IPC.
 * It maintains the MCP server connection and forwards requests from CLI invocations.
 */

import { existsSync, mkdirSync, unlinkSync, writeFileSync, readFileSync } from 'node:fs';
import { dirname } from 'node:path';
import {
  type ServerConfig,
  debug,
  getDaemonTimeoutMs,
  getSocketPath,
  getPidPath,
  getSocketDir,
  getConfigHash,
} from './config.js';
import { connectToServer, listTools, callTool, type ConnectedClient } from './client.js';

// ============================================================================
// Types
// ============================================================================

export interface DaemonRequest {
  id: string;
  type: 'listTools' | 'callTool' | 'ping' | 'close' | 'getInstructions';
  toolName?: string;
  args?: Record<string, unknown>;
}

export interface DaemonResponse {
  id: string;
  success: boolean;
  data?: unknown;
  error?: { code: string; message: string };
}

interface PidFileContent {
  pid: number;
  configHash: string;
  startedAt: string;
}

// ============================================================================
// PID File Management
// ============================================================================

/**
 * Write PID file with config hash for stale detection
 */
export function writePidFile(serverName: string, configHash: string): void {
  const pidPath = getPidPath(serverName);
  const dir = dirname(pidPath);

  if (!existsSync(dir)) {
    mkdirSync(dir, { recursive: true, mode: 0o700 });
  }

  const content: PidFileContent = {
    pid: process.pid,
    configHash,
    startedAt: new Date().toISOString(),
  };

  writeFileSync(pidPath, JSON.stringify(content), { mode: 0o600 });
}

/**
 * Read PID file content
 */
export function readPidFile(serverName: string): PidFileContent | null {
  const pidPath = getPidPath(serverName);

  if (!existsSync(pidPath)) {
    return null;
  }

  try {
    const content = readFileSync(pidPath, 'utf-8');
    return JSON.parse(content);
  } catch {
    return null;
  }
}

/**
 * Remove PID file
 */
export function removePidFile(serverName: string): void {
  const pidPath = getPidPath(serverName);
  try {
    if (existsSync(pidPath)) {
      unlinkSync(pidPath);
    }
  } catch {
    // Ignore errors during cleanup
  }
}

/**
 * Remove socket file
 */
export function removeSocketFile(serverName: string): void {
  const socketPath = getSocketPath(serverName);
  try {
    if (existsSync(socketPath)) {
      unlinkSync(socketPath);
    }
  } catch {
    // Ignore errors during cleanup
  }
}

/**
 * Check if a process is running
 */
export function isProcessRunning(pid: number): boolean {
  try {
    process.kill(pid, 0);
    return true;
  } catch {
    return false;
  }
}

/**
 * Kill a process by PID
 */
export function killProcess(pid: number): boolean {
  try {
    process.kill(pid, 'SIGTERM');
    return true;
  } catch {
    return false;
  }
}

// ============================================================================
// Daemon Worker
// ============================================================================

/**
 * Main daemon entry point - run as detached background process
 */
export async function runDaemon(serverName: string, config: ServerConfig): Promise<void> {
  const socketPath = getSocketPath(serverName);
  const configHash = getConfigHash(config);
  const timeoutMs = getDaemonTimeoutMs();

  let idleTimer: ReturnType<typeof setTimeout> | null = null;
  let mcpClient: ConnectedClient | null = null;
  let server: ReturnType<typeof Bun.listen> | null = null;
  // biome-ignore lint: Socket type from Bun.listen handlers
  const activeConnections = new Set<unknown>();

  // Cleanup function
  const cleanup = async () => {
    debug(`[daemon:${serverName}] Shutting down...`);

    if (idleTimer) {
      clearTimeout(idleTimer);
      idleTimer = null;
    }

    // Close all active socket connections
    for (const conn of activeConnections) {
      try {
        (conn as { end: () => void }).end();
      } catch {
        // Ignore
      }
    }
    activeConnections.clear();

    // Close MCP connection
    if (mcpClient) {
      try {
        await mcpClient.close();
      } catch {
        // Ignore
      }
      mcpClient = null;
    }

    // Close socket server
    if (server) {
      try {
        server.stop();
      } catch {
        // Ignore
      }
      server = null;
    }

    // Clean up files
    removeSocketFile(serverName);
    removePidFile(serverName);

    debug(`[daemon:${serverName}] Cleanup complete`);
  };

  // Reset idle timer
  const resetIdleTimer = () => {
    if (idleTimer) {
      clearTimeout(idleTimer);
    }
    idleTimer = setTimeout(async () => {
      debug(`[daemon:${serverName}] Idle timeout reached, shutting down`);
      await cleanup();
      process.exit(0);
    }, timeoutMs);
  };

  // Handle signals
  process.on('SIGTERM', async () => {
    await cleanup();
    process.exit(0);
  });

  process.on('SIGINT', async () => {
    await cleanup();
    process.exit(0);
  });

  // Ensure socket dir exists
  const socketDir = getSocketDir();
  if (!existsSync(socketDir)) {
    mkdirSync(socketDir, { recursive: true, mode: 0o700 });
  }

  // Remove stale socket if exists
  removeSocketFile(serverName);

  // Write PID file
  writePidFile(serverName, configHash);

  // Connect to MCP server
  try {
    debug(`[daemon:${serverName}] Connecting to MCP server...`);
    mcpClient = await connectToServer(serverName, config);
    debug(`[daemon:${serverName}] Connected to MCP server`);
  } catch (error) {
    console.error(`[daemon:${serverName}] Failed to connect:`, (error as Error).message);
    await cleanup();
    process.exit(1);
  }

  // Handle incoming request
  const handleRequest = async (data: Buffer): Promise<DaemonResponse> => {
    resetIdleTimer();

    let request: DaemonRequest;
    try {
      request = JSON.parse(data.toString());
    } catch {
      return {
        id: 'unknown',
        success: false,
        error: { code: 'INVALID_REQUEST', message: 'Invalid JSON' },
      };
    }

    debug(`[daemon:${serverName}] Request: ${request.type} (${request.id})`);

    if (!mcpClient) {
      return {
        id: request.id,
        success: false,
        error: { code: 'NOT_CONNECTED', message: 'MCP client not connected' },
      };
    }

    try {
      switch (request.type) {
        case 'ping':
          return { id: request.id, success: true, data: 'pong' };

        case 'listTools': {
          const tools = await listTools(mcpClient.client);
          return { id: request.id, success: true, data: tools };
        }

        case 'callTool': {
          if (!request.toolName) {
            return {
              id: request.id,
              success: false,
              error: { code: 'MISSING_TOOL', message: 'toolName required' },
            };
          }
          const result = await callTool(mcpClient.client, request.toolName, request.args ?? {});
          return { id: request.id, success: true, data: result };
        }

        case 'getInstructions': {
          const instructions = mcpClient.client.getInstructions();
          return { id: request.id, success: true, data: instructions };
        }

        case 'close':
          // Graceful shutdown requested
          setTimeout(async () => {
            await cleanup();
            process.exit(0);
          }, 100);
          return { id: request.id, success: true, data: 'closing' };

        default:
          return {
            id: request.id,
            success: false,
            error: { code: 'UNKNOWN_TYPE', message: `Unknown request type: ${request.type}` },
          };
      }
    } catch (error) {
      const err = error as Error;
      return {
        id: request.id,
        success: false,
        error: { code: 'EXECUTION_ERROR', message: err.message },
      };
    }
  };

  // Start Unix socket server
  try {
    server = Bun.listen({
      unix: socketPath,
      socket: {
        open(socket) {
          activeConnections.add(socket);
          debug(`[daemon:${serverName}] Client connected`);
        },
        async data(socket, data) {
          const response = await handleRequest(data);
          socket.write(JSON.stringify(response) + '\n');
        },
        close(socket) {
          activeConnections.delete(socket);
          debug(`[daemon:${serverName}] Client disconnected`);
        },
        error(socket, error) {
          debug(`[daemon:${serverName}] Socket error: ${error.message}`);
          activeConnections.delete(socket);
        },
      },
    });

    debug(`[daemon:${serverName}] Listening on ${socketPath}`);

    // Start idle timer
    resetIdleTimer();

    // Signal readiness by writing to stdout (parent will read this)
    console.log('DAEMON_READY');

  } catch (error) {
    console.error(`[daemon:${serverName}] Failed to start socket server:`, (error as Error).message);
    await cleanup();
    process.exit(1);
  }
}

// ============================================================================
// Entry point when run directly
// ============================================================================

// Check if running as daemon process
if (process.argv[2] === '--daemon') {
  const serverName = process.argv[3];
  const configJson = process.argv[4];

  if (!serverName || !configJson) {
    console.error('Usage: daemon.ts --daemon <serverName> <configJson>');
    process.exit(1);
  }

  let config: ServerConfig;
  try {
    config = JSON.parse(configJson);
  } catch {
    console.error('Invalid config JSON');
    process.exit(1);
  }

  runDaemon(serverName, config).catch((error) => {
    console.error('Daemon failed:', error);
    process.exit(1);
  });
}

File: src/daemon.rs
use crate::config::{
    debug, get_config_hash, get_daemon_timeout_ms, get_pid_path, get_socket_dir, get_socket_path,
    ServerConfig,
};
use crate::errors::{CliError, ErrorCode};
use std::fs;
#[cfg(windows)]
use std::hash::{Hash, Hasher};
use std::sync::Arc;
use std::time::Duration;
use tokio::io::{AsyncReadExt, AsyncWriteExt};
#[cfg(windows)]
use tokio::net::TcpListener;
#[cfg(unix)]
use tokio::net::UnixListener;
use tokio::sync::mpsc;

// ============================================================================
// Types
// ============================================================================

#[derive(serde::Serialize, serde::Deserialize, Debug, Clone)]
#[serde(rename_all = "camelCase")]
pub struct DaemonRequest {
    pub id: String,
    #[serde(rename = "type")]
    pub req_type: String,
    pub tool_name: Option<String>,
    pub args: Option<serde_json::Value>,
}

#[derive(serde::Serialize, serde::Deserialize, Debug, Clone)]
pub struct DaemonResponse {
    pub id: String,
    pub success: bool,
    pub data: Option<serde_json::Value>,
    pub error: Option<DaemonErrorDetail>,
}

#[derive(serde::Serialize, serde::Deserialize, Debug, Clone)]
pub struct DaemonErrorDetail {
    pub code: String,
    pub message: String,
}

#[derive(serde::Serialize, serde::Deserialize, Debug, Clone)]
#[serde(rename_all = "camelCase")]
pub struct PidFileContent {
    pub pid: i32,
    pub config_hash: String,
    pub started_at: String,
}

// ============================================================================
// PID File Management
// ============================================================================

pub fn write_pid_file(server_name: &str, config_hash: &str) {
    let pid_path = get_pid_path(server_name);
    let dir = pid_path.parent().unwrap();

    if !dir.exists() {
        let _ = fs::create_dir_all(dir);
        #[cfg(unix)]
        {
            use std::os::unix::fs::PermissionsExt;
            let _ = fs::set_permissions(dir, fs::Permissions::from_mode(0o700));
        }
    }

    let content = PidFileContent {
        pid: std::process::id() as i32,
        config_hash: config_hash.to_string(),
        started_at: chrono::Utc::now().to_rfc3339(),
    };

    if let Ok(serialized) = serde_json::to_string(&content) {
        let _ = fs::write(&pid_path, serialized);
        #[cfg(unix)]
        {
            use std::os::unix::fs::PermissionsExt;
            let _ = fs::set_permissions(&pid_path, fs::Permissions::from_mode(0o600));
        }
    }
}

pub fn read_pid_file(server_name: &str) -> Option<PidFileContent> {
    let pid_path = get_pid_path(server_name);
    if !pid_path.exists() {
        return None;
    }

    let content = fs::read_to_string(pid_path).ok()?;
    serde_json::from_str(&content).ok()
}

pub fn remove_pid_file(server_name: &str) {
    let pid_path = get_pid_path(server_name);
    let _ = fs::remove_file(pid_path);
}

pub fn remove_socket_file(server_name: &str) {
    let socket_path = get_socket_path(server_name);
    let _ = fs::remove_file(socket_path);
}

#[cfg(windows)]
pub fn get_daemon_addr(server_name: &str, config_hash: &str) -> String {
    let mut hasher = std::collections::hash_map::DefaultHasher::new();
    server_name.hash(&mut hasher);
    config_hash.hash(&mut hasher);
    let port = 40000 + (hasher.finish() % 20000);
    format!("127.0.0.1:{}", port)
}

#[cfg(unix)]
pub fn is_process_running(pid: i32) -> bool {
    unsafe { libc::kill(pid, 0) == 0 }
}

#[cfg(not(unix))]
pub fn is_process_running(_pid: i32) -> bool {
    true
}

#[cfg(unix)]
pub fn kill_process(pid: i32) -> bool {
    unsafe { libc::kill(pid, libc::SIGTERM) == 0 }
}

#[cfg(not(unix))]
pub fn kill_process(_pid: i32) -> bool {
    false
}

// ============================================================================
// Daemon Connection Handler
// ============================================================================

async fn handle_request(
    data: &[u8],
    mcp_connection: &crate::client::McpConnection,
    shutdown_tx: &mpsc::Sender<()>,
) -> DaemonResponse {
    let req: DaemonRequest = match serde_json::from_slice(data) {
        Ok(r) => r,
        Err(_) => {
            return DaemonResponse {
                id: "unknown".to_string(),
                success: false,
                data: None,
                error: Some(DaemonErrorDetail {
                    code: "INVALID_REQUEST".to_string(),
                    message: "Invalid JSON".to_string(),
                }),
            };
        }
    };

    debug(&format!("Daemon Request: {} ({})", req.req_type, req.id));

    let result = match req.req_type.as_str() {
        "ping" => Ok(serde_json::json!("pong")),
        "listTools" => match mcp_connection.list_tools().await {
            Ok(tools) => serde_json::to_value(tools).map_err(|e| e.to_string()),
            Err(e) => Err(e.message),
        },
        "callTool" => {
            if let Some(ref name) = req.tool_name {
                let args = req.args.clone().unwrap_or(serde_json::json!({}));
                match mcp_connection.call_tool(name, args).await {
                    Ok(res) => Ok(res),
                    Err(e) => Err(e.message),
                }
            } else {
                Err("toolName required".to_string())
            }
        }
        "getInstructions" => match mcp_connection.get_instructions().await {
            Ok(inst) => Ok(serde_json::json!(inst)),
            Err(e) => Err(e.message),
        },
        "close" => {
            let tx = shutdown_tx.clone();
            tokio::spawn(async move {
                tokio::time::sleep(Duration::from_millis(100)).await;
                let _ = tx.send(()).await;
            });
            Ok(serde_json::json!("closing"))
        }
        _ => Err(format!("Unknown request type: {}", req.req_type)),
    };

    match result {
        Ok(val) => DaemonResponse {
            id: req.id,
            success: true,
            data: Some(val),
            error: None,
        },
        Err(err) => DaemonResponse {
            id: req.id,
            success: false,
            data: None,
            error: Some(DaemonErrorDetail {
                code: "EXECUTION_ERROR".to_string(),
                message: err,
            }),
        },
    }
}

// ============================================================================
// Daemon Loop
// ============================================================================

pub async fn run_daemon(server_name: &str, config: ServerConfig) -> Result<(), CliError> {
    let socket_path = get_socket_path(server_name);
    let config_hash = get_config_hash(&config);
    let timeout_ms = get_daemon_timeout_ms();

    // 1. Ensure socket dir exists
    let socket_dir = get_socket_dir();
    if !socket_dir.exists() {
        let _ = fs::create_dir_all(&socket_dir);
        #[cfg(unix)]
        {
            use std::os::unix::fs::PermissionsExt;
            let _ = fs::set_permissions(&socket_dir, fs::Permissions::from_mode(0o700));
        }
    }

    // 2. Cleanup stale files
    remove_socket_file(server_name);

    // 3. Connect to MCP server
    debug(&format!(
        "[daemon:{}] Connecting to MCP server...",
        server_name
    ));

    // Direct connection without daemon caching
    let mcp_conn = match &config {
        ServerConfig::Stdio(sc) => {
            let client = crate::client::StdioClient::connect(server_name, sc).await?;
            crate::client::McpConnection::Stdio(client)
        }
        ServerConfig::Http(hc) => {
            let client = crate::client::HttpClient::connect(server_name, hc).await?;
            crate::client::McpConnection::Http(client)
        }
    };
    debug(&format!("[daemon:{}] Connected to MCP server", server_name));

    // 4. Start daemon listener
    #[cfg(unix)]
    let listener = UnixListener::bind(&socket_path).map_err(|e| crate::errors::CliError {
        code: ErrorCode::ClientError,
        error_type: "DAEMON_BIND_FAILED".to_string(),
        message: format!("Failed to bind Unix socket: {}", e),
        details: None,
        suggestion: Some("Check socket directory permissions".to_string()),
    })?;

    #[cfg(windows)]
    let listener = TcpListener::bind(get_daemon_addr(server_name, &config_hash))
        .await
        .map_err(|e| crate::errors::CliError {
            code: ErrorCode::ClientError,
            error_type: "DAEMON_BIND_FAILED".to_string(),
            message: format!("Failed to bind daemon TCP listener: {}", e),
            details: None,
            suggestion: Some("Check whether the daemon port is already in use".to_string()),
        })?;

    // 5. Write PID File
    write_pid_file(server_name, &config_hash);

    // 6. Signal Parent: DAEMON_READY
    println!("DAEMON_READY");

    // Channels for signals and connection processing
    let (shutdown_tx, mut shutdown_rx) = mpsc::channel::<()>(1);
    let (activity_tx, mut activity_rx) = mpsc::channel::<()>(100);

    let mcp_conn_arc = Arc::new(mcp_conn);

    // background loop for socket connections
    let mcp_conn_clone = Arc::clone(&mcp_conn_arc);
    let shutdown_tx_clone = shutdown_tx.clone();
    tokio::spawn(async move {
        loop {
            match listener.accept().await {
                Ok((mut stream, _)) => {
                    let mcp_conn_ref = Arc::clone(&mcp_conn_clone);
                    let sht_tx = shutdown_tx_clone.clone();
                    let act_tx = activity_tx.clone();

                    tokio::spawn(async move {
                        let _ = act_tx.send(()).await; // Signal activity to reset idle timer

                        let mut buffer = vec![0u8; 65536];
                        match stream.read(&mut buffer).await {
                            Ok(n) => {
                                if n > 0 {
                                    let response =
                                        handle_request(&buffer[..n], &mcp_conn_ref, &sht_tx).await;
                                    if let Ok(serialized) = serde_json::to_string(&response) {
                                        let _ = stream.write_all(serialized.as_bytes()).await;
                                        let _ = stream.write_all(b"\n").await;
                                        let _ = stream.flush().await;
                                    }
                                }
                            }
                            Err(e) => {
                                debug(&format!("Socket connection read error: {}", e));
                            }
                        }
                    });
                }
                Err(e) => {
                    debug(&format!("Socket accept error: {}", e));
                }
            }
        }
    });

    // Idle Timer Management
    let idle_duration = Duration::from_millis(timeout_ms);
    let mut idle_timer = tokio::time::interval(idle_duration);
    idle_timer.reset();

    loop {
        tokio::select! {
            _ = shutdown_rx.recv() => {
                debug(&format!("[daemon:{}] Graceful shutdown requested", server_name));
                break;
            }
            _ = activity_rx.recv() => {
                // Reset idle timer
                idle_timer.reset();
            }
            _ = idle_timer.tick() => {
                // Since tick fires immediately on reset sometimes, we verify elapsed
                debug(&format!("[daemon:{}] Idle timeout reached, shutting down", server_name));
                break;
            }
            _ = tokio::signal::ctrl_c() => {
                debug(&format!("[daemon:{}] SIGINT received, shutting down", server_name));
                break;
            }
        }
    }

    // Cleanup
    remove_socket_file(server_name);
    remove_pid_file(server_name);
    if let Ok(conn) = Arc::try_unwrap(mcp_conn_arc) {
        let _ = conn.close().await;
    }
    debug(&format!("[daemon:{}] Cleanup complete", server_name));

    Ok(())
}
